### ISSUE TITLE
opal_convertor_unpack only works if count==1 ?

### ISSUE THREAD

**Issue by alex--m** [open-mpi/ompi#8183]
## TL;DR

For "Generic" datatypes (not the predefined MPI_INT and co.) the unpacker seems to ignore the "count" and only unpack one item from the buffer. 

## Background information

This is an issue I encountered during MCA component development, so it doesn't show on the OMPI API, as far as I can tell. Specifically, I used `MPI_Type_create_struct()` to create a custom datatype, and used `opal_convertor_unpack` to unpack it on the receiver-side. It worked for count=1, but crashes for count=2 (see below). I thought I might have done something wrong, but if I look at `opal_unpack_general_function` I see it practically disregards `pConvertor->count` - and as a result, after 'count == 1' we get to label `complete_loop` (which is probably wrong).

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
latest master

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
manual build from git master.

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.
````
$ git submodule status
 1bd7a3577f64ea796db59f3b13449ef56ff5ddad 3rd-party/openpmix (v1.1.3-2544-g1bd7a357)
 70f69b27435ac98a85e05e5e95e7f47a376b4d37 3rd-party/prrte (dev-30723-g70f69b2743)
````

## Details of the problem

The problem is hard to reproduce, because I'm using internal OMPI functionality.

I've used the small tester (attached), and on ranks #1 and #2 I get the following:
````
dt_test: opal_datatype_unpack.c:388: opal_generic_simple_unpack: Assertion `iov_len_local < opal_datatype_basicDatatypes[pElem->elem.common.type]->size' failed.

Thread 1 "dt_test" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
51      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffffec508b1 in __GI_abort () at abort.c:79
#2  0x00007ffffec4042a in __assert_fail_base (fmt=0x7ffffedc7a38 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x7ffffe9d9ca8 "iov_len_local < opal_datatype_basicDatatypes[pElem->elem.common.type]->size", file=file@entry=0x7ffffe9d9918 "opal_datatype_unpack.c", 
    line=line@entry=388, function=function@entry=0x7ffffe9d9e40 <__PRETTY_FUNCTION__.4> "opal_generic_simple_unpack") at assert.c:92
#3  0x00007ffffec404a2 in __GI___assert_fail (assertion=0x7ffffe9d9ca8 "iov_len_local < opal_datatype_basicDatatypes[pElem->elem.common.type]->size", file=0x7ffffe9d9918 "opal_datatype_unpack.c", 
    line=388, function=0x7ffffe9d9e40 <__PRETTY_FUNCTION__.4> "opal_generic_simple_unpack") at assert.c:101
#4  0x00007ffffe97ca32 in opal_generic_simple_unpack (pConvertor=0x85abf30, iov=0x7ffffffec290, out_size=0x7ffffffec2a0, max_data=0x7ffffffec150) at opal_datatype_unpack.c:388
#5  0x00007ffffe969984 in opal_convertor_unpack (pConv=0x85abf30, iov=0x7ffffffec290, out_size=0x7ffffffec2a0, max_data=0x7ffffffec150) at opal_convertor.c:324
#6  0x00007fffef7a8fea in coll_ucx_generic_datatype_unpack (state=0x85abed0, offset=0, src=0x7ffff318b1d8, length=84) at coll_ucx_datatype.c:115
#7  0x00007ffff2642ac5 in ucg_builtin_comp_unpack ...
(gdb) f 4
#4  0x00007ffffe97ca32 in opal_generic_simple_unpack (pConvertor=0x85abf30, iov=0x7ffffffec290, out_size=0x7ffffffec2a0, max_data=0x7ffffffec150) at opal_datatype_unpack.c:388
388                 assert( iov_len_local < opal_datatype_basicDatatypes[pElem->elem.common.type]->size );
(gdb) p pElem->elem.common.type
$1 = 1
(gdb) p pElem->elem.common.flags 
$2 = 0
(gdb) p posDesc
No symbol "posDesc" in current context.
(gdb) p pos_desc
$3 = 3
(gdb) p pStack->index 
$4 = -1
(gdb) p pElem
$6 = (dt_elem_desc_t *) 0x8567510
(gdb) p *pElem
$7 = {elem = {common = {flags = 0, type = 1}, count = 3, blocklen = 8316293038636793855, extent = 21, disp = 0}, loop = {common = {flags = 0, type = 1}, items = 3, loops = 4294967295, unused = 21, 
    extent = 0}, end_loop = {common = {flags = 0, type = 1}, items = 3, unused = 4294967295, size = 21, first_elem_disp = 0}}
(gdb) p description[0]
$8 = {elem = {common = {flags = 310, type = 4}, count = 1, blocklen = 1, extent = 1, disp = 0}, loop = {common = {flags = 310, type = 4}, items = 1, loops = 1, unused = 1, extent = 0}, end_loop = {
    common = {flags = 310, type = 4}, items = 1, unused = 1, size = 1, first_elem_disp = 0}}
(gdb) p description[1]
$9 = {elem = {common = {flags = 310, type = 16}, count = 1, blocklen = 2, extent = 16, disp = 8}, loop = {common = {flags = 310, type = 16}, items = 1, loops = 2, unused = 16, extent = 8}, 
  end_loop = {common = {flags = 310, type = 16}, items = 1, unused = 2, size = 16, first_elem_disp = 8}}
(gdb) p description[2]
$10 = {elem = {common = {flags = 310, type = 6}, count = 1, blocklen = 1, extent = 4, disp = 28}, loop = {common = {flags = 310, type = 6}, items = 1, loops = 1, unused = 4, extent = 28}, 
  end_loop = {common = {flags = 310, type = 6}, items = 1, unused = 1, size = 4, first_elem_disp = 28}}
(gdb) p description[3]
$11 = {elem = {common = {flags = 0, type = 1}, count = 3, blocklen = 8316293038636793855, extent = 21, disp = 0}, loop = {common = {flags = 0, type = 1}, items = 3, loops = 4294967295, unused = 21, 
    extent = 0}, end_loop = {common = {flags = 0, type = 1}, items = 3, unused = 4294967295, size = 21, first_elem_disp = 0}}
(gdb) p description[4]
$12 = {elem = {common = {flags = 24931, type = 8302}, count = 1667327346, blocklen = 8246126567762239592, extent = 2338324178067088239, disp = 7142842523992485481}, loop = {common = {flags = 24931, 
      type = 8302}, items = 1667327346, loops = 1818304616, unused = 2338324178067088239, extent = 7142842523992485481}, end_loop = {common = {flags = 24931, type = 8302}, items = 1667327346, 
    unused = 1818304616, size = 2338324178067088239, first_elem_disp = 7142842523992485481}}
(gdb) p pElem->elem.common.flags 
$13 = 0
(gdb) p pConvertor 
$14 = (opal_convertor_t *) 0x85abf30
(gdb) p *pConvertor
$15 = {super = {obj_magic_id = 16046253926196952813, obj_class = 0x7ffffec03ea0 <opal_convertor_t_class>, obj_reference_count = 1, cls_init_file_name = 0x7fffef7ad40b "coll_ucx_datatype.c", 
    cls_init_lineno = 260}, remoteArch = 4291367424, flags = 537526532, local_size = 42, remote_size = 42, pDesc = 0x851ff80, use_desc = 0x8520040, count = 2, stack_size = 5, 
  pBaseBuf = 0x7ffffffec820 "", pStack = 0x85abfd8, fAdvance = 0x7ffffe97c0db <opal_generic_simple_unpack>, master = 0x847d8f0, stack_pos = 0, partial_length = 0, bConverted = 0, checksum = 0, 
  csum_ui1 = 0, csum_ui2 = 0, static_stack = {{index = -1, type = 0, padding = 0, count = 0, disp = 32}, {index = 0, type = 4, padding = 0, count = 1, disp = 0}, {index = 0, type = 0, padding = 0, 
      count = 0, disp = 0}, {index = 0, type = 0, padding = 0, count = 0, disp = 0}, {index = 0, type = 0, padding = 0, count = 0, disp = 0}}}
(gdb) 
(gdb) p total_unpacked 
$23 = 0
(gdb) p iov_len_local
$24 = 42
(gdb) up
#5  0x00007ffffe969984 in opal_convertor_unpack (pConv=0x85abf30, iov=0x7ffffffec290, out_size=0x7ffffffec2a0, max_data=0x7ffffffec150) at opal_convertor.c:324
324         return pConv->fAdvance( pConv, iov, out_size, max_data );
(gdb) p pConv
$25 = (opal_convertor_t *) 0x85abf30
(gdb) p pConv->count
$26 = 2
````

This last line of GDB output shows that the logic in `opal_convertor_unpack` seems to stop after one item, and if `pConv->count` is more than 1 - it crashes because it treats the second item as excess data. This seems to contradict the logic in `opal_convertor_prepare_for_recv` and other functions using `opal_convertor_unpack`, because those explicitly prepare a convertor for any `count` items to be unpacked:
````
int32_t opal_convertor_prepare_for_recv( opal_convertor_t* convertor,
                                         const struct opal_datatype_t* datatype,
                                         size_t count,
                                         const void* pUserBuf )
{
    /* Here I should check that the data is not overlapping */

    convertor->flags |= CONVERTOR_RECV;
#if OPAL_CUDA_SUPPORT
    if (!( convertor->flags & CONVERTOR_SKIP_CUDA_INIT )) {
        mca_cuda_convertor_init(convertor, pUserBuf);
    }
#endif

    assert(! (convertor->flags & CONVERTOR_SEND));
    OPAL_CONVERTOR_PREPARE( convertor, datatype, count, pUserBuf );
````

## What next?

Two options: 

1. I can change my code to include `for(i=0; i<count; i++) { opal_convertor_unpack(); }` - but this doesn't look like what the developers of `opal_convertor_unpack` intended.

2. I can change `opal_convertor_unpack` to explicitly loop for any `pConv->count`.

What say yee?

@bosilca @ggouaillardet - tagged you because you seem to be the last co-authors of that area of the code.

**Comment by alex--m** [comment]
Test I used:
````
#include <mpi.h>
#include <stdio.h>
#include <unistd.h>

int main(int argc, char** argv) {
    // Initialize the MPI environment
    MPI_Init(NULL, NULL);

    // Get the number of processes
    int world_size;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    // Get the rank of the process
    int world_rank;
    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);


    struct object {
        char c;
        double x[2];
        float dummy; /* This is NOT included in the MPI datatype */
        int i;
    };

    MPI_Datatype newstructuretype;
#define structlen 3
    int blocklengths[structlen];
    MPI_Datatype types[structlen];
    MPI_Aint displacements[structlen];

    struct object myobject[2] = {0};
    myobject[1].i = world_rank + 1;

    // one character
    blocklengths[0]  = 1;
    types[0]         = MPI_CHAR;
    displacements[0] = offsetof(struct object, c);

    // two doubles
    blocklengths[1]  = 2;
    types[1]         = MPI_DOUBLE;
    displacements[1] = offsetof(struct object, x);

    // one int
    blocklengths[2]  = 1;
    types[2]         = MPI_INT;
    displacements[2] = offsetof(struct object, i);

    MPI_Type_create_struct(structlen, blocklengths, displacements, types,
                           &newstructuretype);
    MPI_Type_commit(&newstructuretype);

    printf("#%i - MPI_Type_create_struct() -> %p (unpacked: %lu)\n",
           getpid(), newstructuretype, sizeof(myobject));

    MPI_Bcast(&myobject, 2, newstructuretype, 0, MPI_COMM_WORLD);

    MPI_Type_free(&newstructuretype);

    if (myobject[1].i != 1) {
        printf("#%i - RANK #%i RESULT IS WRONG! (%p has %i, expected 1)\n",
               getpid(), world_rank, &myobject, myobject[1].i);
    }

    // Finalize the MPI environment.
    MPI_Finalize();
}
````

**Comment by alex--m** [comment]
As I did further debugging - looks like this problem doesn't happen in other flows leading to `opal_convertor_unpack`... I'm closing this for now, until I'm again certain the problem is not in my code.

**Comment by alex--m** [comment]
Problem "solved" - I found that other callers pass just the size of a single datatype as `iov[0].iov_len`, whereas I passed the total size (in my case 2 times the size of the datatype, because `count == 2`).

**Comment by bosilca** [comment]
@alex--m let me give you a quick answer here. The opal_convertor_pack and _unpack use 2 things to do their job, the data description and a stack. There is a trick in there: each datatype ends with a special END_LOOP marker, that has no matching LOOP in the datatype, but instead is meant to be matched with the level 0 in the conversion stack. Thus, every time these functions hit the if branch comparing the type of the element with END_LOOP, they are basically repeating some of the operations they already did, and in case the stack_pos is 0, they will start a new iteration on the user provided count. Ping me outside here if you want more detailed information.
